lib/ip.go

```diff
@@ -36,13 +37,56 @@ func RandIP4List(n int) []net.IP {
 	return ips
 }
 
+// RandIP4Range returns a list of randomly generated IPv4 addresses within
+// starting and ending IPrange.
+
+// note: starting and ending IPs should be in format of "1.1.1.1"
+func RandIP4Range(startIP, EndIP string) (net.IP, error) {
+	StartIPRaw := net.ParseIP(startIP).To4()
+	if len(StartIPRaw) == 0 || len(StartIPRaw) > net.IPv4len {
+		return nil, fmt.Errorf("range is Invalid")
+	}
+	StartIPInt := binary.BigEndian.Uint32(StartIPRaw.To4())
+	EndIPRaw := net.ParseIP(EndIP).To4()
+	if len(EndIPRaw) == 0 || len(EndIPRaw) > net.IPv4len {
+		return nil, fmt.Errorf("range is Invalid")
+	}
+	EndIPInt := binary.BigEndian.Uint32(EndIPRaw.To4())
+	if StartIPInt > EndIPInt {
+		return nil, fmt.Errorf("range is Invalid")
+	}
+	rangeIP := binary.BigEndian.Uint32(RandIP4().To4())
+	temp := EndIPInt - StartIPInt
+	if temp <= 0 {
+		return nil, fmt.Errorf("range is Invalid")
+	}
+	rangeIP %= (EndIPInt - StartIPInt)
+	rangeIP += StartIPInt
+	rangeIPbyte := [4]byte{0, 0, 0, 0}
+	binary.BigEndian.PutUint32(rangeIPbyte[:], rangeIP)
+	return net.IP(rangeIPbyte[:]), nil
+}
+
 // RandIP4ListWrite prints a list of new randomly generated IPv4 addresses.
 func RandIP4ListWrite(n int) {
 	for i := 0; i < n; i++ {
 		fmt.Println(RandIP4())
 	}
 }
 
+// RandIP4ListWrite prints a list of new randomly generated IPv4 addresses
+// within starting and IPs ending range.
+func RandIP4RangeListWrite(min, max string, n int) error {
+	for i := 0; i < n; i++ {
+		ip, err := RandIP4Range(min, max)
```

Comment:
There is some serious slowness in this API call because it has to calculate everything over and over for `min` and `max`, even though these variables are the same for each iteration of the loop. This is even more expensive in the IPv6 case because we make use of `big.Int` math (i.e. multi-precision math), which is inherently costly.

I would suggest that we refactor the logic so that we avoid having to do this. One possible route is to pull out the logic that does all the work on `min` and `max` (i.e. ensure that they are correct ranges, convert them to integers, etc), and make a separate function that assumes valid `min` and `max` integers have been given, so that it goes directly to generating the random IP and adjusting it to fit the given range. This function itself would then call the first function to generate appropriate `min` & `max` integers and do that error checking early, then simply call the final, random-IP generating function over & over in this loop.

The same exact strategy can be applied to the IPv6 code.